packages/k8s.contrib.crd/internal/ModuleGenerator.pkl

```diff
@@ -0,0 +1,366 @@
+local function generatePklClass(schema: JsonSchema, className: Type, typeNames: Type.TypeNames): ClassNode =
+  new {
+    docComment = getDocComment(schema, "class")
+    classHeader {
+      name {
+        value = className.name
+      }
+    }
+    properties = generateClassBody(schema, typeNames)
+  }
+
+function isClassLike(schema: JsonSchema.Schema): Boolean =
+  if (schema is Boolean)
+    false
+  else
+    schema.properties != null
+
+/// Determine the name of a type.
+///
+/// Try to use the parent property's name as part of the class name in case of conflict.
+/// If already at the root, add a number at the end.
+local function determineTypeName(path: List<String>, candidateName: String, existingTypeNames: Set<Type>, index: Int): Type =
+  if (existingTypeNames.contains(utils.pascalCase(candidateName)))
+    if (path.isEmpty)
+      determineTypeName(path, candidateName + index.toString(), existingTypeNames, index + 1)
+    else
+      determineTypeName(path.dropLast(1), utils.pascalCase(path.last.capitalize()) + utils.pascalCase(candidateName), existingTypeNames, index)
+  else
+    new { name = utils.pascalCase(candidateName); moduleName = module.moduleName }
+
+/// The schemas that should be rendered as classes.
+///
+/// Classes get rendered for any subschema that has [JsonSchema.properties] defined, and does not show up in converters
+local classSchemas: Type.TypeNames =
+  utils._findMatchingSubSchemas(filteredRootSchema, List(), (elem) -> elem != filteredRootSchema && isClassLike(elem))
+    .filter((path, _) -> !pathPrefixes(path).any((prefix) -> converters.containsKey(prefix))) // path or prefix are not explicitly in converters
+    .entries
+    .fold(Map(), (accumulator: Type.TypeNames, pair) ->
+    let (path = pair.first)
+      let (schema = pair.second)
+        let (typeName = determineTypeName(path, path.lastOrNull?.capitalize() ?? "Item", accumulator.values.toSet(), 0))
+          accumulator.put(schema, typeName)
+  )
+
+// Return all possible path prefixes of this path, starting with its first entry and ending with the full path
+local function pathPrefixes(path: List<String>): List<List<String>> = new Listing {
+  for (i, _ in path) {
+    path.take(i + 1)
+  }
+}.toList()
+
+// mapSchema deeply transforms a CRD schema. CRD schemas are significantly simplified, as many fields must not be set.
+// Nested types are *always* represented in properties, items or additionalProperties, even if they are also set in oneOf etc
+function mapSchema(
+  schema: JsonSchema.Schema,
+  path: Listing<String>,
+  f: (JsonSchema.Schema) -> JsonSchema.Schema
+): JsonSchema.Schema = let (_ = if (path.length > 32) trace("long path \(path)") else null) let (applied = f.apply(schema)) (applied) {
+    properties = if (applied.properties == null) null else new {
+      for (name, property in applied.properties!!) {
+        [name] = mapSchema(property, (path) { name }, f)
+      }
+    }
```

Comment:
typical `when` case [throughout]
```suggestion
    when (applied.properties != null) {
      properties = new {
        for (name, property in applied.properties!!) {
          [name] = mapSchema(property, (path) { name }, f)
        }
      }
    }
```

Notice that the `if (parent.prop == null) null else`-pattern is not a 0-transaction on `prop`; it deletes the default (which would be `Null(new ExpectedType {})`).